src/systems/systems/src/pneumatic/valve.rs

```diff
@@ -0,0 +1,690 @@
+}
+impl DefaultValve {
+    fn new(open_amount: Ratio) -> Self {
+        Self {
+            open_amount,
+            connector: PneumaticContainerConnector::new(),
+        }
+    }
+
+    pub fn new_closed() -> Self {
+        DefaultValve::new(Ratio::new::<ratio>(0.))
+    }
+
+    pub fn new_open() -> Self {
+        DefaultValve::new(Ratio::new::<ratio>(1.))
+    }
+
+    pub fn open_amount(&self) -> Ratio {
+        self.open_amount
+    }
+
+    pub fn update_move_fluid(
+        &mut self,
+        context: &UpdateContext,
+        container_one: &mut impl PneumaticContainer,
+        container_two: &mut impl PneumaticContainer,
+    ) {
+        self.connector
+            .with_transfer_speed_factor(self.open_amount)
+            .update_move_fluid(context, container_one, container_two);
+    }
+
+    pub fn fluid_flow(&self) -> VolumeRate {
+        self.connector.fluid_flow()
+    }
+}
+impl ControllablePneumaticValve for DefaultValve {
+    fn update_open_amount<T: PneumaticValveSignal, U: ControllerSignal<T>>(
+        &mut self,
+        controller: &U,
+    ) {
+        if let Some(signal) = controller.signal() {
+            self.open_amount = signal.target_open_amount();
+        }
+    }
+}
+impl SimulationElement for DefaultValve {
+    fn accept<T: SimulationElementVisitor>(&mut self, visitor: &mut T) {
+        visitor.visit(self);
+    }
+}
+
+pub struct PneumaticContainerConnector {
+    fluid_flow: VolumeRate,
+    transfer_speed_factor: Ratio,
+}
+impl PneumaticContainerConnector {
+    const HEAT_CAPACITY_RATIO: f64 = 1.4;
+    const TRANSFER_SPEED: f64 = 5.;
+    const HEAT_TRANSFER_SPEED: f64 = 3.;
+
+    pub fn new() -> Self {
+        Self {
+            fluid_flow: VolumeRate::new::<cubic_meter_per_second>(0.),
+            transfer_speed_factor: Ratio::new::<ratio>(1.),
+        }
+    }
+
+    /// Transfer fluid between two containers based on the pressure difference
+    pub fn update_move_fluid(
+        &mut self,
+        context: &UpdateContext,
+        container_one: &mut impl PneumaticContainer,
+        container_two: &mut impl PneumaticContainer,
+    ) {
+        self.heat_conduction(context, container_one, container_two);
+
+        let equalization_volume: Volume = (container_one
+            .pressure()
+            .get::<pascal>()
+            .powf(1. / Self::HEAT_CAPACITY_RATIO)
+            - container_two
+                .pressure()
+                .get::<pascal>()
+                .powf(1. / Self::HEAT_CAPACITY_RATIO))
+            * container_one.volume()
+            * container_two.volume()
+            / (container_two
+                .pressure()
+                .get::<pascal>()
+                .powf(1. / Self::HEAT_CAPACITY_RATIO)
+                * container_one.volume()
+                + container_one
+                    .pressure()
+                    .get::<pascal>()
+                    .powf(1. / Self::HEAT_CAPACITY_RATIO)
+                    * container_two.volume());
```

Comment:
**suggestion**: There's some repetition of calculations here. Maybe you can extract this into a few variables?